ai_eval/coach.py

```diff
@@ -0,0 +1,817 @@
+    )
+
+    workspace_history = List(
+        scope=Scope.user_state,
+        default=[],
+    )
+
+    coach_history = List(
+        scope=Scope.user_state,
+        default=[],
+    )
+
+    evaluation_fragments = List(
+        scope=Scope.user_state,
+        default=[],
+    )
+
+    attempts_used = Integer(
+        scope=Scope.user_state,
+        default=0,
+    )
+
+    input_open = Boolean(
+        scope=Scope.user_state,
+        default=True,
+    )
+
+    final_submission = String(
+        scope=Scope.user_state,
+        default="",
+    )
+
+    final_evaluation_markdown = String(
+        scope=Scope.user_state,
+        default="",
+    )
+
+    max_attempts = Integer(
+        display_name=_("Maximum attempts"),
+        help=_("Total attempts a learner is allowed for evaluation"),
+        default=3,
+        scope=Scope.settings,
+    )
+
+    editable_fields = AIEvalXBlock.editable_fields + (
+        "initial_message",
+        "scenario_data",
+        "workspace_title",
+        "coach_title",
+        "intro_text",
+        "character_1_name",
+        "character_1_role",
+        "character_1_prompt",
+        "character_1_avatar",
+        "character_2_name",
+        "character_2_role",
+        "character_2_prompt",
+        "character_2_avatar",
+        "evaluator_prompt",
+        "allow_reset",
+        "blacklist",
+        "max_attempts",
+    )
+
+    def studio_view(self, context):
+        """Render Studio edit view with styling only (no extra wrapper)."""
+        fragment = super().studio_view(context)
+        try:
+            fragment.add_css(self.resource_string("static/css/coach_studio.css"))
+        except Exception:  # pylint: disable=broad-exception-caught
+            pass
+        return fragment
+
+    def _render_template(self, template, **context):
+        return self._jinja_env.from_string(template).render(context)
+
+    def _get_character_data(self, character_index):  # pylint: disable=missing-function-docstring
+        # Hardcoded at 2 characters but extensible.
+        characters = [
+            {
+                "name": self.character_1_name,
+                "role": self.character_1_role,
+                "avatar": self.character_1_avatar,
+                "pane": "workspace",
+            },
+            {
+                "name": self.character_2_name,
+                "role": self.character_2_role,
+                "avatar": self.character_2_avatar,
+                "pane": "coach",
+            },
+        ]
+        return characters[character_index]
+
+    def _ensure_histories(self):
+        """
+        Initialize chat histories.
+        """
+        if getattr(self, "_histories_ready", False):
+            return
+        workspace = list(self.workspace_history or [])
+        coach = list(self.coach_history or [])
+        evaluations = list(getattr(self, "evaluation_fragments", []) or [])
+        legacy = list(self.chat_history or [])
```

Comment:
why such defensive coding here? Aren't these all defined to have default values?


```suggestion
        workspace = self.workspace_history
        coach = self.coach_history
        evaluations = self.evaluation_fragments
        legacy = self.chat_history
```